lxd/network/network_utils_sriov.go

```diff
@@ -103,157 +104,161 @@ func SRIOVGetHostDevicesInUse(s *state.State) (map[string]struct{}, error) {
-			return "", 0, err
+			return -1, "", err
 		}
 
-		// Use next free VF index.
-		for i := sriovNum + 1; i < sriovTotal; i++ {
-			vfListPath := fmt.Sprintf("/sys/class/net/%s/device/virtfn%d/net", parentDev, i)
-			nicName, err = sriovGetFreeVFInterface(reservedDevices, vfListPath, pfDevID, pfDevPort)
+		for _, ent := range ents {
+			// We expect the entry to be a directory for the VF's interface name.
+			if !ent.IsDir() {
+				continue
+			}
+
+			// We can't use this VF interface as it is reserved by another device.
+			_, exists := reservedDevices[ent.Name()]
+			if exists {
+				continue
+			}
+
+			// Get VF dev_port and dev_id values.
+			vfDevPort, err := ioutil.ReadFile(fmt.Sprintf("%s/%s/dev_port", vfListPath, ent.Name()))
 			if err != nil {
-				return "", 0, err
+				return -1, "", err
 			}
 
-			// Found a free VF.
-			if nicName != "" {
-				vfID = i
-				break
+			vfDevID, err := ioutil.ReadFile(fmt.Sprintf("%s/%s/dev_id", vfListPath, ent.Name()))
+			if err != nil {
+				return -1, "", err
 			}
-		}
-	}
 
-	if nicName == "" {
-		return "", 0, fmt.Errorf("All virtual functions on parent device are already in use")
-	}
+			// Skip VFs if they do not relate to the same device and port as the parent PF.
+			// Some card vendors change the device ID for each port.
+			if bytes.Compare(pfDevPort, vfDevPort) != 0 || bytes.Compare(pfDevID, vfDevID) != 0 {
+				continue
+			}
 
-	return nicName, vfID, nil
-}
+			nicName := ent.Name()
 
-// sriovGetFreeVFInterface checks the contents of the VF directory to find a free VF interface name that
-// belongs to the same device and port as the parent. Returns VF interface name or empty string if
-// no free interface found.
-func sriovGetFreeVFInterface(reservedDevices map[string]struct{}, vfListPath string, pfDevID []byte, pfDevPort []byte) (string, error) {
-	ents, err := ioutil.ReadDir(vfListPath)
-	if err != nil {
-		return "", err
-	}
+			iface, err := net.InterfaceByName(nicName)
+			if err != nil {
+				return -1, "", errors.Wrapf(err, "Failed loading interface %q", nicName)
+			}
 
-	for _, ent := range ents {
-		// We can't use this VF interface as it is reserved by another device.
-		_, exists := reservedDevices[ent.Name()]
-		if exists {
-			continue
-		}
+			// Ignore if interface is up (may be in use by another application already).
+			if iface.Flags&net.FlagUp != 0 {
+				continue
+			}
 
-		// Get VF dev_port and dev_id values.
-		vfDevPort, err := ioutil.ReadFile(fmt.Sprintf("%s/%s/dev_port", vfListPath, ent.Name()))
-		if err != nil {
-			return "", err
-		}
+			addresses, err := iface.Addrs()
+			if err != nil {
+				return 0, "", errors.Wrapf(err, "Failed getting interface addresses for %q", nicName)
+			}
 
-		vfDevID, err := ioutil.ReadFile(fmt.Sprintf("%s/%s/dev_id", vfListPath, ent.Name()))
-		if err != nil {
-			return "", err
-		}
+			// Ignore if interface has IP addresses (may be in use by another application already).
+			if len(addresses) > 0 {
```

Comment:
@stgraber the original issue says we should consider non-link local IPs, but it occurred to me that given we also require the interface to be down, it shouldn't have any automatically configured link-local addresses, and therefore if it has any sort of address on it we should consider it in use, does that ok to you?

Reply:
yeah, sounds good